drivers/fpga/fpga_eos_s3.c

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2021 Antmicro <www.antmicro.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <string.h>
+#include <device.h>
+#include <drivers/fpga.h>
+#include "fpga_eos_s3.h"
+
+void eos_s3_fpga_enable_clk(void)
+{
+	CRU->C16_CLK_GATE = C16_CLK_GATE_PATH_0_ON;
+	CRU->C21_CLK_GATE = C21_CLK_GATE_PATH_0_ON;
+	CRU->C09_CLK_GATE = C09_CLK_GATE_PATH_1_ON | C09_CLK_GATE_PATH_2_ON;
+	CRU->C02_CLK_GATE = C02_CLK_GATE_PATH_1_ON;
+}
+
+void eos_s3_fpga_disable_clk(void)
+{
+	CRU->C16_CLK_GATE = C16_CLK_GATE_PATH_0_OFF;
+	CRU->C21_CLK_GATE = C21_CLK_GATE_PATH_0_OFF;
+	CRU->C09_CLK_GATE = C09_CLK_GATE_PATH_1_OFF | C09_CLK_GATE_PATH_2_OFF;
+	CRU->C02_CLK_GATE = C02_CLK_GATE_PATH_1_OFF;
+}
+
+struct quickfeather_fpga_data {
+	char *FPGA_info;
+};
+
+static enum FPGA_status eos_s3_fpga_get_status(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	if (PMU->FB_STATUS == FPGA_STATUS_ACTIVE) {
+		return FPGA_STATUS_ACTIVE;
+	} else
+		return FPGA_STATUS_INACTIVE;
```

Comment:
Couldn't this function's if/else block be substituted with just `return PMU->FB_STATUS`?

Reply:
Unfortunately not, because `FB_STATUS` can take values other than 1 or 0. The FPGA in the EOS S3 has `ACTIVE`, `DEEP_SLEEP`, and `SHUT_DOWN` statuses. In this case, it will be better if we return only the status that can correspond to the programmability of the FPGA. Therefore, we do not return the value directly because the API is supposed to be abstract. The next step will be to add to support to Zynq Ultrascale+, which will also have different register values, and the abstract statuses should stay the same.

Reply:
So you plan to add other status values for EOS S3 later? Because right now [FPGA_status enum](https://github.com/zephyrproject-rtos/zephyr/blob/e7421b2188c38194993376b493ac8eff2082b974/include/drivers/fpga.h#L15-L24) has only two values defined.